### DISCUSSION TITLE
[ImapSyncMigrationHelper] Some tools to use ImapSync for massive migrations and 365 integration

### DISCUSSION THREAD

**Discussion by adrensnyder** [imapsync/imapsync#434] [Show and tell]
For my work, I have to optimize the structure creation of 365 tenants and their migration. I loved ImapSync before get in 365 world, so I've created some scripts to make things a little easier for use it in massive migrations.

I haven't tested the migration to/from Gmail with ImapSync and oAuth2 yet. I believe it should be sufficient to understand how to interface to create the app, similar to what is done with 365. At that point, using Mutt, you retrieve the token, and from there, it should be all the same. I don't manage the Gmail tenant to be able to try it out

I hope they can help others

Project: https://github.com/adrensnyder/ImapSyncMigrationHelper

Thanks

**Comment by gilleslamiral** [comment]
Ciao Adren

Have you tried
https://imapsync.lamiral.info/oauth2/oauth2_office365/README.txt
https://imapsync.lamiral.info/oauth2/oauth2_office365/

**Reply by adrensnyder** [reply]
> Ciao Adren
> 
> Have you tried https://imapsync.lamiral.info/oauth2/oauth2_office365/README.txt https://imapsync.lamiral.info/oauth2/oauth2_office365/

Hello Gilles. Over a year passed after your answer but i had no time to rework the oauth2.0 i'm using, and for now working, with mutt.
The links you had mentioned are dead. Have you a new links to some new procedures updated over mutt script?
For now, using mutt, i can autoupdate the token seconds prior the expire using "at" command with a script that reprogram the renew. If there is a better system i will consider it :)

A month ago i've update some procedures on my GIT like the one that automatize the app creation, with all needed permissions, on Microsoft 365 with graph.

**Reply by gilleslamiral** [reply]
https://imapsync.lamiral.info/oauth2/oauth2_imap/
https://imapsync.lamiral.info/oauth2/oauth2_imap/README_oauth2.txt

**Comment by adrensnyder** [comment]
Hi Gilles :)

It's the first time I've encountered the procedure you've linked

I've been creating an app for each tenant I want to access every time, and deleting it after I've completed the migration. To automate this procedure, I've developed a PowerShell script that does it, only requiring admin credentials.

As soon as possible, I will try your procedure for the Token management for Linux. In my scripts, I extract the time from the auth generated from Mutt and auto-schedule in a loop with AT to get the token synchronized at the right time.

Thanks for your work. Really appreciated! 👍

**Comment by David-Garcia9919** [comment]
I’ve also worked on a few 365 migration projects. For cases where scripting the OAuth2 flow was too time-consuming, I tested a couple of ready-made tools that handle IMAP and Office 365 directly with OAuth2 tokens. One that I tried was EdbMails, which managed the token handling automatically. It’s not open source, but it can be useful if you just need a quick, full-fidelity mailbox transfer.

**Reply by adrensnyder** [reply]
Thanks for the suggestion

With my procedure i use the mutt script for the oauth2 token generation. Also a bash script that automatically renew it when is about to expire. I created also a powershell script, updated with graph, that automatically create the app needed on 365 side and save all the info for mutt script in a json file. So i think is not much time consuming to set up :)

If the transfer is really quick usually i use the procedure for Imap integrated in the exchange migration section. But prefer 1000% imapsync over Microsoft procedures

**Reply by David-Garcia9919** [reply]
Haha, totally get that — ImapSync really gives you fine-grained control compared to Microsoft’s built-in migration tools. The automation you’ve set up with Mutt, Graph, and PowerShell sounds very efficient, especially with the token renewal handled automatically.

In some of my migrations I’ve gone a slightly different route and used EdbMails, which takes care of the OAuth2 handling and IMAP→365 mailbox moves out of the box. It’s not open source like ImapSync, but for cases where time is tight or multiple tenants are involved, it’s been a reliable option.

Still, your PowerShell + JSON flow for Mutt setup is really clever — love that level of automation.

**Reply by adrensnyder** [reply]
I see the EdbMail software but my customers already complains for the normal cost of a migration ahahaha

Prepare tenant, discuss what account/sharebox to create (worst part), create all the accounts and delegate permissions (and also those things are automatized in my git) and manage migration with all the possible problems like some account that need a Pst import 😞, or some mails to have to be manually copied. (yes I am scrupulous. perhaps too much 😆)

If i say "there is also the cost of the software x" i think i'll get shooted ahahha

**Reply by David-Garcia9919** [reply]
Haha, completely understand — clients often hesitate when extra software costs come up 😅

Your scripting workflow makes a lot of sense, especially since you’ve automated so much already. For teams that don’t have that level of setup, EdbMails can help since it automates the Office 365 and IMAP migration process, is fairly cost-effective, and even comes with 24/7 free customer support if something goes wrong mid-migration.

Still, your PowerShell automation for tenant creation and permission handling sounds impressive — that’s the kind of repeatable process that saves hours in larger projects.
